### ISSUE TITLE
GUI: crashing with remove_items_at/add_item_of_type

### ISSUE THREAD

**Issue by white-haired-uncle** [wesnoth/wesnoth#8490]
### Game and System Information

- **Version:** Battle for Wesnoth v1.17.25+dev (88befc29de9-Clean) x86_64
- **Downloaded from:** git
- **Build info:** g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
- **OS:** `Linux spartan 5.15.0-97-generic #107-Ubuntu SMP Wed Feb 7 13:26:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux`

Libraries
=========

Boost:             1.74
Lua:               5.4.6
OpenSSL/libcrypto: 3.0.0b-dev (runtime 3.0.0b-dev)
libcurl:           7.81.0 (runtime 7.81.0)
Cairo:             1.16.0 (runtime 1.16.0)
Pango:             1.50.6 (runtime 1.50.6)
SDL:               2.0.20 (runtime 2.0.20)
SDL_image:         2.0.5 (runtime 2.0.5)
SDL_mixer:         2.0.4 (runtime 2.0.4)

Features
========

Lua console completion:       yes
D-Bus notifications back end: yes

Current video settings
======================

SDL video drivers:        [x11] wayland KMSDRM dummy
Window size:              1920x1011
Game canvas size:         1920x1011
Final render target size: 1920x1011
Screen refresh rate:      60
Screen dpi:               69.87x69.98
Renderer:                 opengl (hw)
Maximum texture size:     16384x16384
VSync:                    off

### Description of the bug

wesnoth aborts in a couple ways when using remove_items_at/add_item_of_type.  Also, I don't believe those methods return the correct value for position.

I probably have multiple issues here.  I apologize for lumping them together, but I don't know enough to separate them properly.  Part of the problem is when trying to set up a simple example I seemed to have stumbled on another.

`wesnoth: src/gui/widgets/generator_private.hpp:736: gui2::grid& gui2::generator<minimum_selection, maximum_selection, my_placement, select_action>::create_item(int, const gui2::builder_grid&, const widget_data&, const std::function<void(gui2::widget&)>&) [with minimum_selection = gui2::policy::minimum_selection::one_item; maximum_selection = gui2::policy::maximum_selection::one_item; my_placement = gui2::policy::placement::independent; select_action = gui2::policy::select_action::show; gui2::widget_data = std::map<std::__cxx11::basic_string<char>, std::map<std::__cxx11::basic_string<char>, t_string> >]: Assertion `index == -1 || static_cast<unsigned>(index) <= items_.size()' failed.`

`wesnoth: src/gui/widgets/widget.cpp:242: virtual void gui2::widget::place(const point&, const point&): Assertion `size.x >= 0' failed.`

**Comment by white-haired-uncle** [comment]
Saving as draft, more data to come...

**Comment by white-haired-uncle** [comment]
The above is an example I created to try to demonstrate an error I encounter in a more complex scenario (which of course isn't happening here, but I ran into different issues shown above in the process).  In that case, the remove worked just fine but add crashed when I tried to use position=2, like this...




```
-- this is how they are created
dialog.unit_info_mp:add_item_of_type("empty_page")
dialog.unit_info_mp:add_item_of_type("unit_page")
-- here I want to reset the unit page
dialog.unit_info_mp:remove_items_at(2,1)
dialog.unit_info_mp:add_item_of_type("unit_page",2)  -- aborts

wesnoth: src/gui/widgets/generator_private.hpp:736: gui2::grid& gui2::generator<minimum_selection, maximum_selection, my_placement, select_action>::create_item(int, const gui2::builder_grid&, const widget_data&, const std::function<void(gui2::widget&)>&) [with minimum_selection = gui2::policy::minimum_selection::one_item; maximum_selection = gui2::policy::maximum_selection::one_item; my_placement = gui2::policy::placement::independent; select_action = gui2::policy::select_action::show; gui2::widget_data = std::map<std::__cxx11::basic_string<char>, std::map<std::__cxx11::basic_string<char>, t_string> >]: Assertion `index == -1 || static_cast<unsigned>(index) <= items_.size()' failed.
Abort(coredump)
```

**Comment by white-haired-uncle** [comment]
Clearly I don't really know what I'm doing here, and I think in trying to figure out the right way to replace a (multi)page I may have run into a couple of bugs.  I'm also concerned that even if I find something that works, it may be only a special case because I'm removing the last item in the list and adding a new one which may work where replacing an arbitrary item may not.  If anyone can describe the correct process (assuming I'm not doing something silly to begin with), I'd appreciate it.

**Comment by CelticMinstrel** [comment]
Above all else, it shouldn't crash.

**Comment by white-haired-uncle** [comment]
Okay, I'm probably mixing multiple issues here (this issue is about the crash, but I also think add_item_of_type may be broke and that's what I'll be discussing here).  I'll leave it to someone who understands this better to decide if there needs to be another/separate issue.

Did some more testing.  It looks like:

1. add_item_of_type always appends, even if you give it a position.
2. add_item_of_type ONLY accepts -1 or 0 as a position, if provided (always appends?).
3. add_item_of_type ALWAYS returns -1 or 0 (if 0 is specified).  [Is it returning pos relative to end of list and not absolute?]
4. remove_item_at(M,N) works (at least for N=1).
5. When you remove an item, the list "collapses" (vs leaving an empty slot).

I did this:

```
                        local item,pos = dialog.unit_info_mp:add_item_of_type("empty_page")
                        chat(pos)
                        local item,pos = dialog.unit_info_mp:add_item_of_type("unit_info_page")
                        chat(pos)
                        local item,pos = dialog.unit_info_mp:add_item_of_type("empty_page")
                        chat(pos)
                        dialog.unit_info_mp:remove_items_at(2,1)
                        local item,pos = dialog.unit_info_mp:add_item_of_type("unit_info_page")
                        chat(pos)

```
and then tried to update a unit_info_page at dialog.unit_info_mp[2], dialog.unit_info_mp[3], and dialog.unit_info_mp[4].  Only[3] was valid.  I believe this confirms my assumptions 4,5 and most of 1.

Re-reading the wiki seems to indicate my initial assumption that position=index was correct, or at least consistent with the wiki.

**Comment by gfgtdf** [comment]
So i looked at the code and it there seems to be quite a bit of stuff missing in the index handling of `add_item_of_type` also its not really clear whether these are c++ (0 based) or lua (1-based) indicies but before we fix this we should probably decide first how it _should_ behave.

In particular: :
- what should  "insert at position 1" mean? I'd guess insert as the first element.
- what should  "insert at position 2" mean? I'd guess insert as the second element.
- what should  "insert at position -1" mean? I'd guess insert as the last element.
- what should  "insert at position 0" mean? Also insert as the  fist element?

**Comment by white-haired-uncle** [comment]
At first pass, I would say they should start from 1.  Today, if I (lua 
user) do

foo:add_item_of_type("gold")

I access the result as foo[1].

So if I want to access foo[1] and I specify a position to 
add_item_of_type, it should be 1, not 0.

0 could be an error, ignored, append, or use first available?

Negative indicies get a little confusing if you allow remove_items_at to 
take a negative position (-1 for add != -1 for remove). And what does 
for example remove(-10,5) mean in a list where where you only added 
items with odd positions (do you remove 5 existing (non-nil) elements, 
or just anything in the contiguous 5 "slots" starting at position -10?


That said, having position as an integer index seems like a pain to me.  
Say I want remove and replace element 2.  As soon as I remove element 2 
the list is re-indexed and now what was 3 is now 2 and I can't add my 
new element into position 2.  And by removing 2 any references to 
elements other than 1 now have to be updated everywhere I use them.

This would make a lot more sense to me:

foo:add_item_of_type("weapon","my_sword")
foo.my_sword.name.label = "Excalibur"   (vs foo[1].name.label = 
"Excalibur")

Of course, it's probably too late for that.  I suppose the work around 
is just to never delete anything and just append new so I only need to 
update one reference if I make a change.

Perhaps a new insert_item_of_type(type,pos)?  So I remove(2,1), then 
insert(x,2) and all of my references are still valid (or valid again if 
you look at it that way)?  Or maybe add already does this, I'd need to 
test that.

**Comment by gfgtdf** [comment]
> That said, having position as an integer index seems like a pain to me.

Hmm i mean you don't have to use it, but i kinda agree as i don't think i have usecases other than to insert it at last position. 

So the new proposal is to only accept positive integers for a certain position or 0 to insert at the end.

> 0 could be an error, ignored, append, or use first available?

I don't see what the difference between append and first available coudl be.


> This would make a lot more sense to me:
foo:add_item_of_type("weapon","my_sword")
foo.my_sword.name.label = "Excalibur"   (vs foo[1].name.label =
"Excalibur")


well you can already use stuff like 
 ```
local page = foo:add_item_of_type("weapon")
page.name.label = "Excalibur"
 ```
or even
 ```
local pages = {}
pages.my_sword = foo:add_item_of_type("weapon")
pages.my_sword.name.label = "Excalibur"
 ```
So i don't really see the advantage of your suggestion (even i would, it would  not really be an option tbh..).

**Comment by white-haired-uncle** [comment]
> > That said, having position as an integer index seems like a pain to me.
> 
> Hmm i mean you don't have to use it, but i kinda agree as i don't think i have usecases other than to insert it at last position.
> 
> So the new proposal is to only accept positive integers for a certain position or 0 to insert at the end.
> 
> > 0 could be an error, ignored, append, or use first available?
> 
> I don't see what the difference between append and first available coudl be.

foo:add_item_of_type("weapon",1)
foo:add_item_of_type("weapon",3)

Now append is 4, first available is 2.  I would assume add_item_of_type("weapon") would return either 2 or 4, and add_item_of_type("weapon",0) would return the other.  

> 
> > This would make a lot more sense to me:
> > foo:add_item_of_type("weapon","my_sword")
> > foo.my_sword.name.label = "Excalibur"   (vs foo[1].name.label =
> > "Excalibur")
> 
> well you can already use stuff like
> 
> ```
> local page = foo:add_item_of_type("weapon")
> page.name.label = "Excalibur"
> ```
> 
> or even
> 
> ```
> local pages = {}
> pages.my_sword = foo:add_item_of_type("weapon")
> pages.my_sword.name.label = "Excalibur"
> ```
> 
> So i don't really see the advantage of your suggestion (even i would, it would not really be an option tbh..).

Ah, thank you.  I see what I was doing wrong now. add_item_of_type returns item,pos.  I was so fixated on pos I totally overlooked item.  Much better.

Probably need to see if I can find something like get_position(item) to get the current pos for things like remove or selected_index=.

**Comment by CelticMinstrel** [comment]
> its not really clear whether these are c++ (0 based) or lua (1-based) indicies

Uhh, what are you talking about? It's a _Lua_ API, obviously they need to be _Lua_ (1-based) indices.

> * what should  "insert at position 1" mean? I'd guess insert as the first element.
> * what should  "insert at position 2" mean? I'd guess insert as the second element.

"insert at position N" means "after the insert, the new element is now at position N". I'm pretty sure the only way to meet that requirement is for any elements after it to be pushed forward.

> * what should  "insert at position -1" mean? I'd guess insert as the last element.

Yes. Also, if there are 5 elements, then inserting at position 6 also makes it the last element. (Inserting at position 7 would be an error.)

In keeping with the above definition, though, I wonder if it makes sense for negative numbers to have slightly different behaviour? If the intention is that "after the insert, the new element is at position N", then con

> * what should  "insert at position 0" mean? Also insert as the  fist element?

Inserting at position 0 is an error. There is no position 0.

---

In keeping with the above definition, though, I wonder if it makes sense for negative numbers to have slightly different behaviour? If the intention is that "after the insert, the new element is at position N", then consider a sample case where there's 4 elements:

```
 1     2     3     4
[A]   [B]   [C]   [D]
-4    -3    -2    -1
```

Now you want to insert a new element, `[Q]` at index -2. The result of doing that should probably be:

```
 1     2     3     4    5
[A]   [B]   [C]   [Q]   [D]
-5    -4    -3    -2    -1
```

In short, inserting at a positive index means "insert _before_ the element at this index", while inserting at a negative index means "insert _after_ the element at this index".

---

> foo:add_item_of_type("weapon",1)
> foo:add_item_of_type("weapon",3)
> 
> Now append is 4, first available is 2. I would assume add_item_of_type("weapon") would return either 2 or 4, and add_item_of_type("weapon",0) would return the other.

This doesn't exist. There's no such thing as a "hole" in the list of elements. If from an empty list you add one at 1 and then add one at 3, either the second add is an error, or the second one ends up at index 2. (I'd lean towards the latter, and it should signal this fact by returning 2.)

**Comment by gfgtdf** [comment]
> Uhh, what are you talking about? It's a Lua API, obviously they need to be Lua (1-based) indices.

I'm talking about the (previous) implementation of add_item_of_type, imo that was pretty clear form what i said.

> Yes. Also, if there are 5 elements, then inserting at position 6 also makes it the last element. (Inserting at position 7 would be an error.)

But why though, is there any other lua api that behaves like that? Or are there any actualy usecases other than having a default value for "insert at end" ?

**Comment by white-haired-uncle** [comment]
On 2024-03-09 11:57, Celtic Minstrel wrote:
>> its not really clear whether these are c++ (0 based) or lua
>> (1-based) indicies
> 
> Uhh, what are you talking about? It's a _Lua_ API, obviously they need
> to be _Lua_ (1-based) indices.
> 
>> * what should "insert at position 1" mean? I'd guess insert as the
>> first element.
>> * what should "insert at position 2" mean? I'd guess insert as the
>> second element.
> 
> "insert at position N" means "after the insert, the new element is now
> at position N". I'm pretty sure the only way to meet that requirement
> is for any elements after it to be pushed forward.
> 
>> * what should "insert at position -1" mean? I'd guess insert as
>> the last element.
> 
> Yes. Also, if there are 5 elements, then inserting at position 6 also
> makes it the last element. (Inserting at position 7 would be an
> error.)
> 
> In keeping with the above definition, though, I wonder if it makes
> sense for negative numbers to have slightly different behaviour? If
> the intention is that "after the insert, the new element is at
> position N", then con
> 
>> * what should "insert at position 0" mean? Also insert as the fist
>> element?
> 
> Inserting at position 0 is an error. There is no position 0.
> -------------------------
> 
> In keeping with the above definition, though, I wonder if it makes
> sense for negative numbers to have slightly different behaviour? If
> the intention is that "after the insert, the new element is at
> position N", then consider a sample case where there's 4 elements:
> 
>  1     2     3     4
> [A]   [B]   [C]   [D]
> -4    -3    -2    -1
> 
> Now you want to insert a new element, [Q] at index -2. The result of
> doing that should probably be:
> 
>  1     2     3     4    5
> [A]   [B]   [C]   [Q]   [D]
> -5    -4    -3    -2    -1
> 
> In short, inserting at a positive index means "insert _before_ the
> element at this index", while inserting at a negative index means
> "insert _after_ the element at this index".
> -------------------------
> 
>> foo:add_item_of_type("weapon",1)
>> foo:add_item_of_type("weapon",3)
>> 
>> Now append is 4, first available is 2. I would assume
>> add_item_of_type("weapon") would return either 2 or 4, and
>> add_item_of_type("weapon",0) would return the other.
> 
> This doesn't exist. There's no such thing as a "hole" in the list of
> elements. If from an empty list you add one at 1 and then add one at
> 3, either the second add is an error, or the second one ends up at
> index 2. (I'd lean towards the latter, and it should signal this fact
> by returning 2.)

Okay, I thought I tested this.  I know I checked that the idea was valid 
with a lua array.  But at the moment it seems any attempt (at least 
those I tried) to specify a position with add_item_of_type drop core.

As far as what to do if I tried to do this, assuming that not allowing 
the user to create the elements out of order is acceptable (fair), I'd 
vote for error.  If I'm doing something wrong, please give me a useful 
error as soon as it can be detected, don't let me get away with it by 
helpfully giving me something that seems to work but is not what I 
explicitly asked for (even when providing the corrected value, which 
I'll probably overlook).  Also, I think you also stated that it should 
be an error above in your example of adding element 7 to a length 5 
list.

**Comment by CelticMinstrel** [comment]
> I know I checked that the idea was valid with a lua array.

Certainly it's possible to punch holes in a Lua array by assigning nil to an element, but the list of elements in a tree view or multipage is not a Lua array. It's a dynamic array in the C++ backend. Removing an element isn't assigning null to it, it's shifting every element after it down one so as to overwrite the element to be removed.

> If I'm doing something wrong, please give me a useful error as soon as it can be detected

I guess this is fair. I'll retract my request that it not be an error to insert past the maximum

**Comment by CelticMinstrel** [comment]
> But why though, is there any other lua api that behaves like that? Or are there any actualy usecases other than having a default value for "insert at end" ?

[table.insert](https://www.lua.org/manual/5.4/manual.html#pdf-table.insert) does exactly that.

**Comment by gfgtdf** [comment]
> [table.insert](https://www.lua.org/manual/5.4/manual.html#pdf-table.insert) does exactly that.

I don't see anything in there that says negative indices refer to counting form the back there

**Comment by CelticMinstrel** [comment]
`table.insert` does exactly the behaviour you actually quoted. You're correct that it doesn't handle negative indices. However, it's my opinion that handling negative indices as counting from the end is just a nice feature so we should include it on those grounds.

But if you're talking about the behaviour of "negative means inserting _after_ instead of _before_", then yeah, I'm unaware of any Lua API that does it. And if you expand to other programming languages… I tried Python, but it doesn't do that – a negative index still inserts before. In my opinion, it seems more intuitive to maintain the invariant "after inserting at position N, the element is at position N" (or in code, `table.insert(list, N, value) assert(list[N] == value`) than it is to maintain the invariant "the new element is inserted before the specified element".

**Comment by white-haired-uncle** [comment]
Fixed with #8516
